inngest/src/main/kotlin/com/inngest/Step.kt

```diff
@@ -100,15 +108,34 @@ class Step(
             }
         } catch (e: StateNotFound) {
             // If there is no existing result, run the lambda
-            val data = fn()
-            throw StepInterruptException(id, hashedId, data)
+            executeStep(id, hashedId, fn)
+        } catch (e: StepError) {
+            throw e
         }
-        // TODO - Catch Step Error here and throw it when error parsing is added to getState
 
         // TODO - handle invalidly stored step types properly
         throw Exception("step state incorrect type")
     }
 
+    private fun <T> executeStep(
+        id: String,
+        hashedId: String,
+        fn: () -> T,
+    ) {
+        try {
+            val data = fn()
+            throw StepInterruptException(id, hashedId, data)
+        } catch (exception: Exception) {
+            when (exception) {
+                is RetryAfterError,
+                is NonRetriableError,
```

Comment:
similar comment as the set of `ResultStatusCode`s, but again not worth doing if other SDKs aren't